File: src/commonutils/hadoop/HdfsUtil.java
package commonutils.hadoop;

import commonutils.text.StringUtil;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.*;
import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Objects;
import java.util.function.Predicate;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class HdfsUtil {
    private String url;
    private String user = "hdfs";
    private FileSystem fs;
    private final Logger logger = LogManager.getLogger(this.getClass());

    public HdfsUtil(String url) {
        assert StringUtil.isNotBlank(url);
        this.url = url;
        init();
    }

    public HdfsUtil(String url, String user) {
        assert StringUtil.isNotBlank(url) && StringUtil.isNotBlank(user);
        this.url = url;
        this.user = user;
        init();
    }

    private void init() {
        assert url.startsWith("hdfs://") || url.startsWith("/");
        Pattern pattern = Pattern.compile("^(hdfs)://(([\\d]{1,3}.[\\d]{1,3}.[\\d]{1,3}.[\\d]{1,3})|([1-9a-zA-Z.]+)):([\\d]{1,5})");
        Matcher matcher = pattern.matcher(url);
        if (matcher.matches()) {
            url = matcher.group(0);
        } else {
            throw new IllegalArgumentException("Base hdfs url is invalid!");
        }
        System.setProperty("HADOOP_USER_NAME", user);
        System.setProperty("user.name", user);
        try {
            this.fs = FileSystem.get(new URI(url), new Configuration(true));
        } catch (IOException e) {
            logger.error("Error when initialize hdfs file system!", e);
            e.printStackTrace();
        } catch (URISyntaxException e) {
            e.printStackTrace();
        }
    }

    public void setUser(String user) {
        this.user = user;
    }

    public ArrayList<String> list(String path) {
        return list(path, (Objects::nonNull));
    }

    public ArrayList<String> listFiles(String path) {
        return list(path, FileStatus::isFile);
    }

    public ArrayList<String> listDirs(String path) {
        return list(path, FileStatus::isDirectory);
    }

    private ArrayList<String> list(String path, Predicate<FileStatus> fileFilter) {
        ArrayList<String> ret = new ArrayList<>();
        if (!exists(path)) {
            return ret;
        }
        if (isFile(path)) {
            ret.add(getPath(path).toUri().getRawPath());
            return ret;
        }
        try {
            return new ArrayList<>(Arrays.asList(fs.listStatus(getPath(path)))).stream().parallel().filter(fileFilter)
                    .map(x -> x.getPath().toUri().getRawPath()).collect(Collectors.toCollection(ArrayList::new));
        } catch (IOException e) {
            logger.error("Error when get file list!", e);
        }
        return ret;
    }

    public ArrayList<String> listAll(String path) {
        return listAll(path, Objects::nonNull);
    }

    public ArrayList<String> listAllFiles(String path) {
        return listAll(path, FileStatus::isFile);
    }

    public ArrayList<String> listAllDirs(String path) {
        return listAll(path, FileStatus::isDirectory);
    }

    private ArrayList<String> listAll(String path, Predicate<FileStatus> fileFilter) {
        ArrayList<String> arrayList = new ArrayList<>();
        if (!exists(path)) {
            return arrayList;
        }
        list(path, fileFilter).forEach(x->{
            if (isDirectory(x)) {
                arrayList.addAll(listAll(x, fileFilter));
            }
            arrayList.add(x);
        });
        return arrayList;
    }

    public boolean delete(String... path) {
        assert null != path && path.length > 0 : "Paths should not be empty or null!";
        boolean ret = true;
        for (String s : path) {
            try {
                ret &= fs.delete(getPath(s), true);
            } catch (IOException e) {
                logger.error("Error when delete " + s, e);
            }
        }
        return ret;
    }

    public boolean exists(String path) {
        boolean exits = false;
        try {
            exits = fs.exists(getPath(path));
        } catch (IOException e) {
            logger.error(e);
        }
        return exits;
    }

    public boolean createFile(String path) {
        boolean created = false;
        if (!exists(path)) {
            try {
                created = fs.createNewFile(getPath(path));
            } catch (IOException e) {
                logger.error("Error when create file " + path, e);
            }
        }
        return created;
    }

    public boolean mkdir(String path) {
        boolean created = false;
        if (!exists(path)) {
            try {
                created = fs.mkdirs(getPath(path));
            } catch (IOException e) {
                logger.error("Error when create file " + path, e);
            }
        }
        return created;
    }

    public boolean chown(String user, String group, String... path) {
        boolean ch = true;
        for (String s : path) {
            try {
                if (StringUtil.isBlank(group)) {
                    fs.setOwner(getPath(s), user, getFileStatus(s).getGroup());
                } else {
                    fs.setOwner(getPath(s), user, group);
                }
            } catch (Exception e) {
                ch = false;
                logger.error("Error when chown " + s, e);
            }
        }
        return ch;
    }

    public boolean chown(String user, String... path) {
        return chown(user, null, path);
    }

    private FileStatus getFileStatus(String path) {
        FileStatus ret = null;
        try {
            ret = fs.getFileStatus(getPath(path));
        } catch (IOException e) {
            logger.error("Error when get file status: " + path, e);
        }
        return ret;
    }

    public boolean rename(String sourcePath, String destPath) {
        boolean ret = false;
        try {
            ret = fs.rename(getPath(sourcePath), getPath(destPath));
        } catch (IOException e) {
            logger.error("Error when rename " + sourcePath + " to " + destPath, e);
        }
        return ret;
    }

    public boolean copy(String sourcePath, String destPath, boolean overwrite) {
        boolean c = false;
        try {
            c = FileUtil.copy(fs, getPath(sourcePath), fs, getPath(destPath), false, overwrite, new Configuration());
        } catch (IOException e) {
            logger.error("Error when copy " + sourcePath + " to " + destPath, e);
        }
        return c;
    }

    public boolean move(String sourcePath, String destPath) {
        return rename(sourcePath, destPath);
    }

    public boolean moveToTrash(String... path) {
        boolean b = true;
        String trashPath = "/user/" + user + "/.Trash/";
        if (!exists(trashPath)) {
            mkdir(trashPath);
        }
        for (String s : path) {
            b &= move(s, trashPath);
        }
        return b;
    }

    public boolean chmod(String path, String mod) {
        boolean b = false;
        try {
            b = FileUtil.chmod(path, mod, true) == 0;
        } catch (IOException e) {
            logger.error("Error when chmod " + path, e);
        }
        return b;
    }

    public boolean put(String sourcePath, String destPath) {
        boolean b = false;
        try {
            fs.copyFromLocalFile(getPath(sourcePath), getPath(destPath));
            b = true;
        } catch (IOException e) {
            logger.error("Error when copy from " + sourcePath + " to " + destPath, e);
        }
        return b;
    }

    public boolean get(String sourcePath, String destPath) {
        boolean b = false;
        try {
            fs.copyToLocalFile(getPath(sourcePath), getPath(destPath));
            b = true;
        } catch (IOException e) {
            logger.error("Error when copy from " + sourcePath + " to " + destPath, e);
        }
        return b;
    }

    public boolean isFile(String path) {
        return getFileStatus(path).isFile();
    }

    public boolean isDirectory(String path) {
        return getFileStatus(path).isDirectory();
    }

    private Path getPath(String path) {
        return new Path(path);
    }

}

File: src/scala/com/nam/bigdata/HDFSUtil.scala
package scala.com.nam.bigdata
import java.net.URI

import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.permission.{FsAction, FsPermission}
import org.apache.hadoop.fs.{FileStatus, FileSystem, FileUtil, Path}
import org.apache.log4j.Logger

import scala.collection.mutable.ArrayBuffer

class HDFSUtil(hdfsBasePath: String) {

  private[this] val logger = Logger.getLogger(this.getClass)

  System.setProperty("HADOOP_USER_NAME", "hdfs")
  System.setProperty("user.name", "hdfs")
  private val fsUri = new URI(hdfsBasePath)
  private val fs = FileSystem.get(fsUri, new Configuration())

  //设置HDFS用户
  def setUser(user: String): Unit = {
    System.setProperty("HADOOP_USER_NAME", user)
    System.setProperty("user.name", user)
  }

  /**
    * 列出目录下所有文件,不包括目录
    *
    * @param path
    * @return
    */
  def listFiles(path: String): Array[String] = {
    fs.listStatus(getPath(path)).map(_.getPath.toUri.getPath)
  }

  /**
    * 列出目录下所有目录
    *
    * @param path
    * @param recursive
    * @return
    */
  def listDirs(path: String, recursive: Boolean = false): Array[String] = {
    val arr = ArrayBuffer[String]()
    val paths = fs.listStatus(getPath(path)).filter(_.isDirectory).map(_.getPath.toUri.getPath)
    if (recursive) {
      paths.foreach { x =>
        listDirs(x, recursive).foreach(arr.append(_))
        arr.append(x)
      }
    } else {
      paths.foreach(arr.append(_))
    }
    arr.toArray
  }

  /**
    * 列出最底层目录
    *
    * @param path
    * @return
    */
  def listBottomDirs(path: String): Array[String] = {
    val dirs = listDirs(path, true)
    dirs.filter { x =>
      listDirs(x, true).length == 0
    }
  }

  /**
    * 列出所有目录
    *
    * @param path      目录
    * @param recursive 递归
    * @return
    */
  def listAllFiles(path: String, recursive: Boolean): Array[String] = {
    val arr = ArrayBuffer[String]()
    val files = fs.listFiles(getPath(path), recursive)
    while (files.hasNext) {
      val f = files.next()
      if (f.isFile)
        arr.append(f.getPath.toUri.getPath)
    }
    arr.toArray
  }

  /**
    * 删除
    *
    * @param paths
    */
  def delete(paths: String*): Unit = {
    paths.par.foreach { path =>
      if (fs.exists(getPath(path))) {
        if (fs.isDirectory(getPath(path))) {
          deleteDir(path)
        } else {
          deleteFile(path)
        }
      } else {
        logger.error(s"$path not exists, SKIP DELETE!")
      }
    }
  }

  /**
    * 删除目录
    *
    * @param paths
    */
  private def deleteDir(paths: String*): Unit = {
    paths.par.foreach { path =>
      if (fs.exists(getPath(path)) && fs.isDirectory(getPath(path))) {
        logger.info(s"try to delete $path")
        val dirs = listDirs(path)
        deleteDir(dirs: _*)
        val files = listAllFiles(path, recursive = false)
        deleteFile(files: _*)
        fs.delete(getPath(path), true)
      } else {
        logger.error(s"$path is not a Dir or not exits")
      }
    }
  }

  /**
    * 删除文件
    *
    * @param paths
    */
  private def deleteFile(paths: String*): Unit = {
    paths.par.foreach { path =>
      if (fs.exists(getPath(path))) {
        if (fs.isFile(getPath(path))) {
          logger.info(s"try to delete $path")
          val res = fs.delete(getPath(path), true)
          if (res) {
            logger.info(s"complete deleting $path")
          } else {
            logger.info(s"an error occurred during deleting $path")
          }
        } else {
          logger.warn(s"$path is not a file! SKIP DELETE!")
        }
      } else {
        logger.warn(s"$path not exits! SKIP DELETE!")
      }
    }
  }

  /**
    * 删除到回收站
    *
    * @param paths
    */
  def moveToTrash(paths: String*): Unit = {
    val hadoop_user_name = System.getProperty("HADOOP_USER_NAME").trim
    val user_name = System.getProperty("user.name").trim
    val user = if (hadoop_user_name.isEmpty) {
      if (user_name.isEmpty) {
        "hdfs"
      } else {
        user_name
      }
    } else {
      hadoop_user_name
    }
    val trashPath = s"/user/$user/.Trash"
    paths.par.foreach { path =>
      move(path, trashPath)
    }
  }

  /**
    * 移动
    *
    * @param source
    * @param dest
    */
  def move(source: String, dest: String): Unit
  = {
    if (!exists(source)) {
      logger.error(s"Sourc path:$source not exists!")
    } else {
      copy(source, dest, deleteSourceFile = true)
      logger.info(s"moved $source to $dest!")
    }
  }

  /**
    * 复制
    *
    * @param source           源地址
    * @param destination      目标地址
    * @param deleteSourceFile 是否删除源文件
    * @param overWrite        是否覆盖
    * @return
    */
  def copy(source: String, destination: String, deleteSourceFile: Boolean = false, overWrite: Boolean = false): Unit = {
    val sourceFile = getPath(source)
    val destFile = getPath(destination)

    if (fs.exists(sourceFile)) {
      val sourceDir = new FileStatus()
      sourceDir.setPath(sourceFile)
      if (!fs.exists(destFile)) {
        FileUtil.copy(fs, sourceDir, fs, destFile, deleteSourceFile, overWrite, fs.getConf)
      } else {
        if (fs.isDirectory(sourceFile) && fs.isFile(destFile)) {
          logger.error("can't copy a directory to a file!", new RuntimeException("can't copy a directory to a file!"))
        } else if (fs.isFile(sourceFile) && fs.isFile(destFile)) {
          logger.error(s"destination:$destination already exists!", new RuntimeException(s"destination:$destination already exists!"))
        } else {
          FileUtil.copy(fs, sourceDir, fs, destFile, deleteSourceFile, overWrite, fs.getConf)
        }
      }
    } else {
      logger.error(s"source:$source not exists!", new RuntimeException(s"source:$source not exists!"))
    }

  }

  /**
    * 获取HDFS地址
    *
    * @param path
    * @return
    */
  def getPath(path: String): Path = {
    //    if (path.toLowerCase().startsWith("hdfs://")) {
    //      new Path(path)
    //    } else {
    //      new Path(hdfsBasePath + path)
    //    }
    new Path(path)
  }

  /**
    * 改变所属用户
    *
    * @param path      路径
    * @param group     所属组
    * @param owner     所属用户
    * @param recursive 是否递归
    */
  def chown(path: String, group: String, owner: String, recursive: Boolean = true): Unit = {
    val p = getPath(path)
    val oldGroup = fs.getFileStatus(p).getGroup
    val oldOwner = fs.getFileStatus(p).getOwner

    if (group.equals(oldGroup) && owner.equals(oldOwner)) {
      logger.warn("组名和用户名未改变")
    } else {
      try {
        fs.setOwner(p, owner, group)
        if (recursive && fs.getFileStatus(p).isDirectory) {
          listDirs(path).par.foreach(p1 => chown(p1, group, owner, recursive))
          listAllFiles(path, recursive = false).par.foreach(p1 => chown(p1, group, owner, recursive = false))
        }
        logger.info(s"changed owner for [$path]")
      } catch {
        case e: Exception =>
          e.printStackTrace()
      }
    }
  }

  /**
    * 改变权限
    *
    * 0 无
    * 1 可执行
    * 2 写
    * 4 读
    * 3 可执行 + 写
    * 5 可执行 + 读
    * 6 读 + 写
    * 7 读 + 写 + 可执行
    *
    * @param path            地址
    * @param userPermission  所属用户权限
    * @param groupPermission 所属组权限
    * @param otherPermission 其他用户权限
    * @param recursive       是否递归
    */
  def chmod(path: String, userPermission: Int, groupPermission: Int, otherPermission: Int, recursive: Boolean): Unit = {

    require(userPermission <= 7 && userPermission > 0
      && groupPermission <= 7 && groupPermission > 0
      && otherPermission <= 7 && otherPermission > 0, "Invalid permission!")

    /**
      * 根据数字获取权限信息
      *
      * @param p
      * @return
      */
    def getPermission(p: Int): FsAction = {
      p match {
        case 1 => FsAction.EXECUTE
        case 2 => FsAction.WRITE
        case 3 => FsAction.WRITE_EXECUTE
        case 4 => FsAction.READ
        case 5 => FsAction.READ_EXECUTE
        case 6 => FsAction.READ_WRITE
        case 7 => FsAction.ALL
        case _ => FsAction.NONE
      }
    }

    if (fs.isDirectory(getPath(path)) && recursive) {
      val dirs = listDirs(path)
      if (dirs.length > 0) {
        dirs.par.foreach { p =>
          chmod(p, userPermission, groupPermission, otherPermission, recursive)
        }
      }
      val files = listAllFiles(path, recursive = false)
      if (files.length > 0) {
        files.par.foreach { p =>
          chmod(p, userPermission, groupPermission, otherPermission, recursive)
        }
      }
    }

    val m: FsPermission = new FsPermission(getPermission(userPermission), getPermission(groupPermission), getPermission(otherPermission))

    fs.setPermission(getPath(path), m)
  }

  /**
    * 改变权限
    *
    * @param path      路径
    * @param mod       权限 形如 "777", "766", "664" 的形式
    * @param recursive 递归
    */
  def chmod(path: String, mod: String, recursive: Boolean = true): Unit = {
    require(mod.length > 0 && mod.length <= 3, s"Invalid permission number, required 3 but found ${mod.length}!")
    mod.foreach { x =>
      val c = x.toInt
      require(c >= 48 && c <= 55, s"Invalid permisson at number $x")
    }
    val arr = mod.split("")
    val u = arr(0).toInt
    val g = arr(1).toInt
    val o = arr(2).toInt
    chmod(path, u, g, o, recursive)
  }

  /**
    * 创建目录
    *
    * @param path
    * @return
    */
  def mkDir(path: String): Unit = {
    if (fs.exists(getPath(path))) {
      if (fs.isDirectory(getPath(path))) {
        logger.error(s"$path already exits")
      } else {
        logger.warn(s"already has a file named $path! but we created a directory.")
        fs.mkdirs(getPath(path))
      }
    } else {
      fs.mkdirs(getPath(path))
    }
  }

  /**
    * 创建文件
    *
    * @param path
    * @return
    */
  def createFile(path: String): Unit = {
    if (fs.exists(getPath(path))) {
      if (fs.isFile(getPath(path))) {
        logger.error(s"$path already exits")
      } else {
        fs.createNewFile(getPath(path))
      }
    } else {
      fs.createNewFile(getPath(path))
    }
  }

  /**
    * 重命名
    *
    * @param path1 原地址
    * @param path2 目标地址
    */
  def rename(path1: String, path2: String): Unit = {
    fs.rename(getPath(path1), getPath(path2))
  }

  /**
    * 地址是否存在
    *
    * @param path
    * @return
    */
  def exists(path: String): Boolean = {
    fs.exists(getPath(path))
  }

  def close: Unit = {
    fs.close()
  }



}